torch/csrc/autograd/saved_variable.cpp

```diff
@@ -1,49 +1,61 @@
 #include "torch/csrc/autograd/saved_variable.h"
 
 #include "torch/csrc/autograd/function.h"
+#include "torch/csrc/autograd/variable.h"
+#include "torch/csrc/autograd/edge.h"
+#include "torch/csrc/jit/tracer_state.h"
 
-using namespace at;
+#include <ATen/Tensor.h>
 
-namespace torch { namespace autograd {
+#include <cstdint>
+#include <list>
+#include <memory>
 
-SavedVariable::SavedVariable(const Variable& variable, bool is_output)
-  : SavedVariable() {
+namespace torch {
+namespace autograd {
+
+SavedVariable::SavedVariable(const Variable& variable, bool is_output) {
   if (!variable.defined()) {
     return;
   }
-  data = variable.data();
-  requires_grad = variable.requires_grad();
-  expected_version = variable.current_version();
-  version = variable.get()->version_counter.save();
-  has_grad_fn = !variable.is_leaf();
-  output_nr = variable.output_nr();
-  if (!has_grad_fn) {
-    grad_accumulator = variable.grad_accumulator();
+  was_default_constructed_ = false;
+
+  data_ = variable.data();
+  requires_grad_ = variable.requires_grad();
+  version_counter_ = variable.version_counter();
+  saved_version_ = version_counter_.current_version();
+  has_grad_fn_ = !variable.is_leaf();
+  output_nr_ = variable.output_nr();
+  if (!has_grad_fn_) {
+    grad_accumulator_ = variable.grad_accumulator();
   }
   if (!is_output) {
-    _grad_fn = variable.grad_fn();
+    grad_fn_ = variable.grad_fn();
   }
-  if (variable.tracing_state()) {
-    tracing_state.reset(new jit::tracer::ValueTracingState(*variable.tracing_state()));
+  if (variable.has_tracing_state()) {
+    tracing_state_.reset(
+        new jit::tracer::ValueTracingState(variable.tracing_state()));
   }
 }
 
-auto SavedVariable::unpack(std::shared_ptr<Function> saved_for) const -> Variable {
-  if (!data.defined()) {
-    if (version.defined()) {
+SavedVariable::~SavedVariable() = default;
+
+Variable SavedVariable::unpack(std::shared_ptr<Function> saved_for) const {
+  if (!data_.defined()) {
+    if (!was_default_constructed_) {
```

Comment:
What was wrong with using `version` to check this? Why did we loose `SavedVersion`? It was a nice API that encapsulated the version checking that you have to do manually now

Reply:
I began by trying to untangle the mutual `friend`-ness of `SavedVariable` and `VariableVersion`. For this I had to move the `VariableVersion` pointer out of the `SavedVariable`. At that point the struct became a bit useless. I also did some static analysis in my head to find that `saved_version.defined()` would only be false when the `SavedVariable` is default constructed. It's like 100 lines less code overall so I thought it would be nice.

Reply:
I'm confused by your statement "mutual friend-ness of SavedVariable and VariableVersion". SavedVariable does not have any friends.

Reply:
It was `VariableVersion` and `SavedVersion`, sorry. The two classes that used to be in this file.

Reply:
I'd love to know what the new relationships between Variable, SavedVariable, VariableVersion, et al are.

Reply:
From the viewpoint of the rest of the code the relationships are still the same. I just turned the `SavedVersion` struct into a `uint32_t`. `SavedVersion` previously had a back-reference to the `VariableVersion` so that it could:
1. check if the version it stored differs from the current version of the `VariableVersion` it references. This has been replaced by the `saved_version_ != version_counter_.current_version()` that is now in `SavedVariable` and
2. check if the `SavedVersion` is `defined()` (by checking if the `version_block_` of the `VariableVersion` it references is `nullptr`). Since `VariableVersion` always has a non-`nullptr` `version_block`, the latter would only be the case if `SavedVersion` was default-constructed. In turn, it would only be default-constructed if the `SavedVariable` was default constructed. So I replaced this whole behavior with the `was_default_constructed` boolean.
This was an implementation detail of `SavedVariable` and `VariableVersion`, so the rest of the code will still construct `SavedVariable`s from `Variable`s as before.